### PR TITLE
MEN-4117: Support headless configuration on Raspberry Pi

### PR DESCRIPTION
**Pull request by lluiscampos**
By editing upstream systemd services to expect boot partition in /uboot
instead of /boot

Changelog: raspberrypi_config: Modify headless configuration services to
expect boot partition in /uboot instead of /boot.

### CODE REVIEW

**Comment by lluiscampos** [comment]
Alternative (and simpler) implementation of #319.

File: configs/raspberrypi_config [hunk 1]
```diff
@@ -91,6 +91,12 @@ function platform_modify() {
     run_and_log_cmd "sudo ln -fs /uboot/overlays work/rootfs/boot/overlays"
     run_and_log_cmd "sudo ln -fs /uboot/$CMDLINE work/rootfs/boot/$CMDLINE"
 
+    # Raspberry Pi headless configuration expects boot partition to be mounted
+    # at /boot, so replace these services with /uboot
+    run_and_log_cmd "find work/rootfs/lib/ -type f -name *.service | xargs sed -i.bak 's|/boot/|/uboot/|g'"
+    log_info "Diff from modified service files:\n$(find work/rootfs/lib/ -type f -name *.service | xargs -I@ bash -c 'diff @.bak @')"
+    run_and_log_cmd "find work/rootfs/lib/ -type f -name *.service.bak | xargs rm"
```
Comment (lluiscampos) [inline configs/raspberrypi_config:98]: This is a bit noisy. I am unsure about the value. The output looks like:

```
2021-03-05 11:41:10 [INFO] [mender-convert-modify] Running hook: platform_modify
2021-03-05 11:41:10 [INFO] [mender-convert-modify] Diff from modified service files:
3c3
< ConditionPathExists=/boot/os_config.json
---
> ConditionPathExists=/uboot/os_config.json
2,3c2,3
< Description=Turn on SSH if /boot/ssh is present
< ConditionPathExistsGlob=/boot/ssh{,.txt}
---
> Description=Turn on SSH if /uboot/ssh is present
> ConditionPathExistsGlob=/uboot/ssh{,.txt}
8c8
< ExecStart=/bin/sh -c "systemctl enable --now ssh && rm -f /boot/ssh ; rm -f /boot/ssh.txt"
---
> ExecStart=/bin/sh -c "systemctl enable --now ssh && rm -f /uboot/ssh ; rm -f /uboot/ssh.txt"
3c3
< ConditionPathExists=/boot/wpa_supplicant.conf
---
> ConditionPathExists=/uboot/wpa_supplicant.conf
10c10
< ExecStart=/bin/mv /boot/wpa_supplicant.conf /etc/wpa_supplicant/wpa_supplicant.conf
---
> ExecStart=/bin/mv /uboot/wpa_supplicant.conf /etc/wpa_supplicant/wpa_supplicant.conf
2021-03-05 11:41:10 [INFO] [mender-convert-modify] Running hook: grow_data_partition
...
```

File: configs/raspberrypi_config [hunk 2]
```diff
@@ -91,6 +91,12 @@ function platform_modify() {
     run_and_log_cmd "sudo ln -fs /uboot/overlays work/rootfs/boot/overlays"
     run_and_log_cmd "sudo ln -fs /uboot/$CMDLINE work/rootfs/boot/$CMDLINE"
 
+    # Raspberry Pi headless configuration expects boot partition to be mounted
+    # at /boot, so replace these services with /uboot
+    run_and_log_cmd "find work/rootfs/lib/ -type f -name *.service | xargs sed -i.bak 's|/boot/|/uboot/|g'"
```
Comment (oleorhagen) [inline configs/raspberrypi_config:96]: Hmm, won't you need to quote the `wildcard` here? :thinking:

File: configs/raspberrypi_config (hunk 1 above)
Comment (oleorhagen) [reply configs/raspberrypi_config:98]: What if you do `run_and_log_cmd` for the diff instead? And then add a `log_info`, with something like:
`Certain service files have been changed to align with our /uboot boot partition mount point. See convert.log for more information.`?

File: configs/raspberrypi_config [hunk 3]
```diff
@@ -91,6 +91,12 @@ function platform_modify() {
...
+    log_info "Diff from modified service files:\n$(find work/rootfs/lib/ -type f -name *.service | xargs -I@ bash -c 'diff @.bak @')"
```
Comment (kacf) [inline configs/raspberrypi_config:97]: `diff -u` maybe?

Also, is there a `log_debug`?

File: configs/raspberrypi_config (hunk 1 above)
Comment (kacf) [reply configs/raspberrypi_config:98]: +1 to Ole's suggestion!

**Review by kacf — APPROVED** [review]
Just nitpicks, looks good overall!

File: configs/raspberrypi_config (hunk 1 above)
Comment (lluiscampos) [reply configs/raspberrypi_config:98]: I tried this... the problem is that `convert.log` has disappeared (at least from my runs). But you guys are pointing me to the right direction so I am going to 1) fix whatever issue is there with the file and 2) move the diff to the file only.

File: configs/raspberrypi_config (hunk 3 above)
Comment (lluiscampos) [reply configs/raspberrypi_config:97]: `diff -u` was even more verbose, so I went for the simple one. But as I move the output to the log file I can restore `-u` here.

File: configs/raspberrypi_config (hunk 2 above)
Comment (lluiscampos) [reply configs/raspberrypi_config:96]: Right, good catch. (good future proof)

File: configs/raspberrypi_config (hunk 1 above)
Comment (lluiscampos) [reply configs/raspberrypi_config:98]: The log file was there... I just didn't know where :weary:

**Comment by lluiscampos** [comment]
Reworded now the logs. And new commit to echo the log file (to help short-sighted developers like myself).

stout:
```
...
2021-03-08 10:24:26 [INFO] [mender-convert-modify] Running hook: platform_modify
2021-03-08 10:24:27 [INFO] [mender-convert-modify] Certain service files have been changed to align with our /uboot boot partition mount point. See convert.log for more information
2021-03-08 10:24:27 [INFO] [mender-convert-modify] Running hook: grow_data_partition
...
```

log file:
```
2021-03-08 10:24:27 [DEBUG] [mender-convert-modify] Diff from modified service files:
--- work/rootfs/lib/systemd/system/apply_noobs_os_config.service.bak	2017-07-17 11:46:17.000000000 +0000
+++ work/rootfs/lib/systemd/system/apply_noobs_os_config.service	2021-03-08 10:24:27.417595917 +0000
@@ -1,6 +1,6 @@
 [Unit]
 Description=Apply NOOBS config
-ConditionPathExists=/boot/os_config.json
+ConditionPathExists=/uboot/os_config.json
 Before=display-manager.service getty.target autologin@tty1.service
 
 [Service]
--- work/rootfs/lib/systemd/system/sshswitch.service.bak	2019-07-22 09:07:25.000000000 +0000
+++ work/rootfs/lib/systemd/system/sshswitch.service	2021-03-08 10:24:27.421596017 +0000
@@ -1,11 +1,11 @@
 [Unit]
-Description=Turn on SSH if /boot/ssh is present
-ConditionPathExistsGlob=/boot/ssh{,.txt}
+Description=Turn on SSH if /uboot/ssh is present
+ConditionPathExistsGlob=/uboot/ssh{,.txt}
 After=regenerate_ssh_host_keys.service
 
 [Service]
 Type=oneshot
-ExecStart=/bin/sh -c "systemctl enable --now ssh && rm -f /boot/ssh ; rm -f /boot/ssh.txt"
+ExecStart=/bin/sh -c "systemctl enable --now ssh && rm -f /uboot/ssh ; rm -f /uboot/ssh.txt"
 
 [Install]
 WantedBy=multi-user.target
--- work/rootfs/lib/systemd/system/raspberrypi-net-mods.service.bak	2019-04-30 09:39:13.000000000 +0000
+++ work/rootfs/lib/systemd/system/raspberrypi-net-mods.service	2021-03-08 10:24:27.425596117 +0000
@@ -1,13 +1,13 @@
 [Unit]
 Description=Copy user wpa_supplicant.conf
-ConditionPathExists=/boot/wpa_supplicant.conf
+ConditionPathExists=/uboot/wpa_supplicant.conf
 Before=dhcpcd.service
 After=systemd-rfkill.service
 
 [Service]
 Type=oneshot
 RemainAfterExit=yes
-ExecStart=/bin/mv /boot/wpa_supplicant.conf /etc/wpa_supplicant/wpa_supplicant.conf
+ExecStart=/bin/mv /uboot/wpa_supplicant.conf /etc/wpa_supplicant/wpa_supplicant.conf
 ExecStartPost=/bin/chmod 600 /etc/wpa_supplicant/wpa_supplicant.conf
 ExecStartPost=/usr/sbin/rfkill unblock wifi 

```

**Pull request merged**